src/Processors/Formats/Impl/NpyRowInputFormat.cpp

```diff
@@ -0,0 +1,434 @@
+#include <IO/ReadHelpers.h>
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <type_traits>
+#include <unordered_map>
+#include <Processors/Formats/Impl/NpyRowInputFormat.h>
+#include <Formats/FormatFactory.h>
+#include <Formats/EscapingRuleUtils.h>
+#include <DataTypes/Serializations/SerializationNullable.h>
+#include <DataTypes/DataTypeString.h>
+#include <Common/typeid_cast.h>
+#include <Common/Exception.h>
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnArray.h>
+#include <Columns/ColumnsNumber.h>
+#include <Storages/IStorage.h>
+#include <Columns/IColumn.h>
+#include <Core/Field.h>
+#include <Core/NamesAndTypes.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/IDataType.h>
+#include <DataTypes/Serializations/ISerialization.h>
+#include <IO/ReadBuffer.h>
+#include <IO/WriteHelpers.h>
+#include <Processors/Formats/IRowInputFormat.h>
+#include <base/types.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+    extern const int BAD_ARGUMENTS;
+    extern const int LOGICAL_ERROR;
+}
+
+
+DataTypePtr createDataType(size_t depth, DataTypePtr nested_type)
+{
+    DataTypePtr result_type = nested_type;
+
+    assert(depth > 0);
+    if (depth > 1)
+    {
+        for (size_t i = 0; i < depth - 1; ++i)
+            result_type = std::make_shared<DataTypeArray>(std::move(result_type));
+    }
+    return result_type;
+}
+
+/*
+Checks, in what endian format data was written.
+return -1: if data is written in little-endian;
+
+         1: if data is written in big-endian;
+
+         0: if data is written in no-endian. */
+int endianOrientation(String descr)
+{
+    if (descr.length() < 3)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Descr field length must be bigger or equal 3");
+    if (descr[0] == '<')
+        return -1;
+    else if (descr[0] == '>')
+        return 1;
+    else if (descr[0] == '|')
+        return 0;
+    else
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Wrong content of field descr");
+}
+
+DataTypePtr parseType(String type)
```

Comment:
You only check for types that starts with `<`, which means we don't support big-endian. And I am not sure it's good to use our data types, maybe better to create a special classes for Numpy data types as I described above